xdsl/dialects/arith.py

```diff
@@ -75,19 +82,27 @@ def from_int_constant(val: Union[int, Attribute],
             attributes={"value": IntegerAttr.from_params(val, typ)})
 
 
-@irdl_op_definition
-class Addi(Operation):
-    name: str = "arith.addi"
-    lhs = OperandDef(signlessIntegerLike)
-    rhs = OperandDef(signlessIntegerLike)
-    result = ResultDef(signlessIntegerLike)
+@dataclass
+class BinaryOperation(Operation):
+    """A generic operation. Operation definitions inherit this class."""
 
     # TODO replace with trait
     def verify_(self) -> None:
         if self.lhs.typ != self.rhs.typ or self.rhs.typ != self.result.typ:
             raise VerifyException(
                 "expect all input and result types to be equal")
 
+    def __hash__(self) -> int:
+        return id(self)
+
+
+@irdl_op_definition
+class Addi(BinaryOperation):
+    name: str = "arith.addi"
+    lhs = OperandDef(signlessIntegerLike)
```

Comment:
I think you could move these into the BinaryOperation.

Reply:
We could indeed have a `IntegerBinaryOperation` or something like this.
In the future, I would like having something like `BinaryOperation[signlessIntegerLike]`, though this would require the merging of #210, and then other improvement (so we should not wait to have this).

Reply:
Ah yes, they have a type constraint that isn't the same everywhere, you are right!